perf/src/sigverify.rs

```diff
@@ -505,41 +505,42 @@ impl Deduper {
 
 //inplace shrink a batch of packets
 pub fn shrink_batches(batches: &mut Vec<PacketBatch>) -> usize {
-    let mut i = 0;
-    let mut ii = 0;
-    let mut j = 1;
-    let mut jj = 0;
-    let mut last = 0;
-    while j < batches.len() {
-        while jj < batches[j].packets.len() {
-            if batches[j].packets[jj].meta.discard() {
-                jj = jj.saturating_add(1);
+    let mut keep_batch_index = 0;
+    let mut keep_packet_index = 0;
+    let mut shrunken_length = 0;
+    for batch_index in 0..batches.len() {
+        for packet_index in 0..batches[batch_index].packets.len() {
+            if batches[batch_index].packets[packet_index].meta.discard() {
                 continue;
             }
-            last = j;
-            let mut done = false;
-            while i < j && !done {
-                while ii < batches[i].packets.len() {
-                    if batches[i].packets[ii].meta.discard() {
-                        batches[i].packets[ii] = batches[j].packets[jj].clone();
-                        batches[j].packets[jj].meta.set_discard(true);
-                        last = i;
-                        done = true;
-                        break;
+            'concentrate_kept: while keep_batch_index < batch_index {
+                while keep_packet_index < batches[keep_batch_index].packets.len() {
+                    let this_keep_packet_index = keep_packet_index;
+                    solana_sdk::saturating_add_assign!(keep_packet_index, 1);
+                    let discard = batches[keep_batch_index].packets[this_keep_packet_index]
+                        .meta
+                        .discard();
+                    if discard {
+                        unsafe {
+                            std::ptr::write(
```

Comment:
We can use safely `std::mem::replace()` instead at the cost of an extra copy.

Reply:
did you bench it before and after?

Reply:
id rather not add an unsafe

Reply:
Yeah there was no apparent perf difference with `std::ptr::write`, switching back to the safe version